fdbserver/workloads/SidebandSingle.actor.cpp

```diff
@@ -0,0 +1,245 @@
+		state Transaction tr(cx);
+		loop {
+			try {
+				Optional<Value> val = wait(tr.get(StringRef(format("Sideband/Client/%d", self->clientId))));
+				if (!val.present()) {
+					throw operation_failed();
+				}
+				SidebandInterface sideband;
+				BinaryReader br(val.get(), IncludeVersion());
+				br >> sideband;
+				TraceEvent("SidebandFetched", sideband.id()).detail("ClientIdx", self->clientId);
+				return sideband;
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+	}
+
+	ACTOR Future<Void> mutator(SidebandSingleWorkload* self, Database cx) {
+		state SidebandInterface checker = wait(self->fetchSideband(self, cx));
+		state double lastTime = now();
+		state Version commitVersion;
+		state bool unknown = false;
+
+		loop {
+			wait(poisson(&lastTime, 1.0 / self->operationsPerSecond));
+			state Transaction tr0(cx);
+			state Transaction tr(cx);
+			state uint64_t key = deterministicRandom()->randomUniqueID().hash();
+
+			state Standalone<StringRef> messageKey(format("Sideband/Message/%llx", key));
+			// first set, this is the "old" value, always retry
+			loop {
+				try {
+					tr0.set(messageKey, LiteralStringRef("oldbeef"));
+					wait(tr0.commit());
+					break;
+				} catch (Error& e) {
+					wait(tr0.onError(e));
+				}
+			}
+			// second set, the checker should see this, no retries on unknown result
+			loop {
+				try {
+					// tr.setOption(FDBTransactionOptions::CAUSAL_WRITE_RISKY);
+					tr.set(messageKey, LiteralStringRef("deadbeef"));
+					wait(tr.commit());
+					commitVersion = tr.getCommittedVersion();
+					break;
+				} catch (Error& e) {
+					if (e.code() == error_code_commit_unknown_result) {
+						unknown = true;
+						++self->messages;
+						checker.updates.send(SidebandMessage(key, invalidVersion));
+						break;
+					}
+					wait(tr.onError(e));
+				}
+			}
+			if (unknown) {
+				unknown = false;
+				continue;
+			}
+			++self->messages;
+			checker.updates.send(SidebandMessage(key, commitVersion));
+		}
+	}
+
+	ACTOR Future<Void> checker(SidebandSingleWorkload* self, Database cx) {
+		loop {
+			state SidebandMessage message = waitNext(self->interf.updates.getFuture());
+			state Standalone<StringRef> messageKey(format("Sideband/Message/%llx", message.key));
+			state Transaction tr(cx);
+			loop {
+				try {
+					tr.setOption(FDBTransactionOptions::USE_GRV_CACHE);
+					state Optional<Value> val = wait(tr.get(messageKey));
+					if (!val.present()) {
+						TraceEvent(SevError, "CausalConsistencyError1", self->interf.id())
+						    .detail("MessageKey", messageKey.toString().c_str())
+						    .detail("RemoteCommitVersion", message.commitVersion)
+						    .detail("LocalReadVersion",
+						            tr.getReadVersion().get()); // will assert that ReadVersion is set
+						++self->consistencyErrors;
+					} else if (val.get() != LiteralStringRef("deadbeef")) {
+						// check again without cache, and if it's the same, that's expected
```

Comment:
In what case would you not get `deadbeef`?

Reply:
In the mutator above, `commit_unknown_result` does not get retried, which should mean there is a chance the `deadbeef` did not get written, depending on what happened behind the scenes of the `commit_unknown_result`. The other side of this is to check if the cache is incorrectly using a stale read version, so if it reads `oldbeef` while a regular transaction sees `deadbeef` we know something has gone wrong

Reply:
If this only happens for `commit_unknown_result`, we could also (or instead) confirm that's what happened by checking for `invalidVersion` in the message.